Encouraging Contributors (#2212)
microsoft/STL: CONTRIBUTING.md

> a CLA and decorate the PR appropriately (e.g., status check, comment). Simply follow the instructions

Author: I don't agree. Building and consuming doesn't necessarily mean contribution intention.

> a CLA and decorate the PR appropriately (e.g., status check, comment). Simply follow the instructions

Author: Possibly we can move out "reporting an issue" instead.

> a CLA and decorate the PR appropriately (e.g., status check, comment). Simply follow the instructions

Reviewer: I saw that you created an issue for this!  Thanks :) We can continue discussion there.

provided by the bot. You will only need to do this once across all repos using our CLA.

# Contributions that we appreciate

## Submit a pull request that aligns with our goals and priorities

 * Look for pinned issues
 * Check issues under the labels `high priority` and `help wanted`
 * New contributors may also look into issues labeled `good first issue`

## Review a pull request

> We welcome reviews from contributors! Especially if this review happens before maintainers' review.

Reviewer: Just adding a few sentences to expand on the rationale here :) 
```suggestion
We welcome code reviews from contributors! Reviews from other contributors can often accelerate the reviewing process by helping a PR reach a more finished state before maintainers review the changes.  As a result, such a PR may require fewer maintainer review iterations before reaching a "Ready to Merge" state!  
```

> To get an idea about our Code Review process, see:

Reviewer: ```suggestion
To gain insight into our Code Review process, you may want to view:
```

 * Ongoing reviews in [Pull Requests List](https://github.com/microsoft/STL/pulls)
 * [Advice for Reviewing Pull Requests](https://github.com/microsoft/STL/wiki/Advice-for-Reviewing-Pull-Requests).
 * [Code Review Videos](https://github.com/microsoft/STL/wiki/Code-Review-Videos!)

# PR Checklist

Before submitting a pull request, please ensure that:

* Identifiers in product code changes are properly `_Ugly` as per

>   https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.

Reviewer: Do we often link to https://eel.is/c++draft in official documentation for the repo?  I have seen it cited in PR comments and issues on occasion, but since it's technically an unofficial source...
> (Generated on 2021-11-07 from the LaTeX sources by cxxdraft-htmlgen. This is not an ISO publication.
For historical versions of the document, see Tim Song's cppwp page.)
> Note: this is an early draft. It's known to be incomplet and incorrekt, and it has lots of bad formatting.

...I wonder if we should instead cite a specific working draft?

Author: I think we should go with [[lex.name]/3.1](https://eel.is/c++draft/lex.name#3.1)

Author: We use stable names in source comments and we do use draft references

* These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).

> * Your changes are written from scratch using only this repository, the C++
>   Working Draft (including any cited standards), other WG21 papers (excluding
>   reference implementations outside of proposed standard wording), and LWG
>   issues as reference material. If your changes are derived from a project
>   that's already listed in NOTICE.txt, that's fine, but please mention it.
>   If your changes are derived from any other project, you *must* mention it
>   here, so we can determine whether the license is compatible and what else
>   needs to be done.

Reviewer: Would this be clearer if we split it into a list of acceptable sources vs. unacceptable?  
```suggestion
* Your changes are written from scratch using only acceptable sources: 
   * this repository
   * the C++ Working Draft (including any cited standards)
   * other WG21 papers (excluding
  reference implementations outside of proposed standard wording)
   * LWG issues as reference material
   * a project listed in NOTICE.txt (please cite the project in the PR description)
  If your changes are derived from any other project, you *must* mention it
  here, so we can determine whether the license is compatible and whether any other 
  steps need to be taken
```

a CLA and decorate the PR appropriately (e.g., status check, comment). Simply follow the instructions
provided by the bot. You will only need to do this once across all repos using our CLA.

# Types of Contributions

> ## Submit a pull request that aligns with our goals and priorities

Reviewer 2: These should probably be `##` and `###` respectively, so as to be placed under the `Contributing` \<h1\>.

The STL repo has many filed issues that track work to be completed.
If you're unsure of where to start, you may want to start by:
 * Looking for pinned issues
 * Checking issues under the labels `high priority` and `help wanted`
 * New contributors may also look into issues labeled `good first issue`

## Review a pull request

We welcome code reviews from contributors! Reviews from other contributors can often accelerate the reviewing process
by helping a PR reach a more finished state before maintainers review the changes.  As a result, such a PR may require
fewer maintainer review iterations before reaching a "Ready to Merge" state!
To gain insight into our Code Review process, you may want to view:

>  * Ongoing reviews in [Pull Requests List](https://github.com/microsoft/STL/pulls)

Reviewer 2: ```suggestion
 * pull requests which are [undergoing review][review:changes-requested],
```

(with a link block):

```md
[review:changes-requested]: /microsoft/STL/pulls?q=is%3Apr+is%3Aopen+review%3Achanges-requested
```

Author: I'm not really sure if it should be changes requested state. Maybe review requested state?

>  * [Advice for Reviewing Pull Requests](https://github.com/microsoft/STL/wiki/Advice-for-Reviewing-Pull-Requests).

Reviewer: ```suggestion
 * [Advice for Reviewing Pull Requests](https://github.com/microsoft/STL/wiki/Advice-for-Reviewing-Pull-Requests)
```

>  * [Code Review Videos](https://github.com/microsoft/STL/wiki/Code-Review-Videos!)

Reviewer 2: ```suggestion
 * our [code review videos][].
```

separate link (plus minor grammar)

# PR Checklist

> Before submitting a pull request, please ensure that:

Reviewer 2: ```suggestion
Before submitting a pull request, please ensure:
```

> * Identifiers in product code changes are properly `_Ugly` as per

Reviewer 2: ```suggestion
1. Any non-standard identifiers in product code (including local variables, exposition-only members, etc.)
   begin with an underscore and a capital letter, as per [\[lex.name\]p3.1][]
```

grammar, make the link out-of-line, etc.

Author: Was not able to make it non-inline link

>   [[lex.name]/3.1](https://eel.is/c++draft/lex.name#3.1)
>   or there are no product code changes.

Reviewer 2: ```suggestion
```
no longer necessary.

> * These changes introduce no known ABI breaks (adding members, renaming

Reviewer 2: ```suggestion
2. Your changes don't introduce change ABI for any type or function that a user may use (including adding or removing non-static data
```

renaming a member does not break ABI, but removing one does

>   members, adding virtual functions, changing whether a type is an aggregate

Reviewer 2: ```suggestion
  members, adding or removing virtual member functions, changing whether a type is an aggregate
```

  or trivially copyable, etc.).

> * Your changes are written from scratch using only acceptable sources:

Reviewer 2: ```suggestion
3. Your changes are written from scratch using only acceptable sources:
```

>    * this repository

Reviewer 2: ```suggestion
   * this repository,
```

>    * the C++ Working Draft (including any cited standards)

Reviewer 2: ```suggestion
   * the C++ Working Draft (including any cited standards),
```

   * other WG21 papers (excluding

>   reference implementations outside of proposed standard wording)

Reviewer 2: ```suggestion
  reference implementations outside of proposed standard wording),
```

>    * LWG issues as reference material

Reviewer 2: ```suggestion
   * LWG issues,
```

>    * a project listed in NOTICE.txt (please cite the project in the PR description)

Reviewer 2: ```suggestion
   * a project listed in [NOTICE.txt][] (make sure to cite the project in the PR description!).
```

make sure to add the link block

>   If your changes are derived from any other project, you *must* mention it

Reviewer 2: ```suggestion
  If your changes are derived from any other project, you _must_ mention it
```

I like `_` for italicizing

a CLA and decorate the PR appropriately (e.g., status check, comment). Simply follow the instructions
provided by the bot. You will only need to do this once across all repos using our CLA.

## Types of Contributions

### Submitting a Pull Request
The STL repo has many open issues that track work which needs to be completed.
If you're unsure of where to start, you may want to:
 * look for pinned issues, or
 * check issues under the labels [`good first issue`][label:"good first issue"],
  [`high priority`][label:"high priority"], or [`help wanted`][label:"help wanted"].

### Reviewing a Pull Request

We love code reviews from contributors! Reviews from other contributors can often accelerate the reviewing process
by helping a PR reach a more finished state before maintainers review the changes. As a result, such a PR may require
fewer maintainer review iterations before reaching a "Ready to Merge" state.

To gain insight into our Code Review process, you can check out:
 * pull requests which are [undergoing review][review:changes-requested],

>  * [Advice for Reviewing Pull Requests][wiki:advice-for-reviewing]
>  * [Code Review Videos][wiki:videos]

Reviewer 2: ```suggestion
 * [Advice for Reviewing Pull Requests][wiki:advice-for-reviewing], and
 * [Code Review Videos][wiki:videos].
```

## PR Checklist

Before submitting a pull request, please ensure:

> 1. Any non-standard identifiers in product code (including local variables, exposition-only members, etc.)

Reviewer 2: ```suggestion
1. Any non-standard identifiers in product code (including local variables, exposition-only members, etc.).
```

> 1. Any non-standard identifiers in product code (including local variables, exposition-only members, etc.).

Reviewer 2: ```suggestion
1. Any non-standard identifiers in product code (including local variables, exposition-only members, etc.)
```

>    begin with an underscore and a capital letter, as per [[lex.name]/3.1](https://eel.is/c++draft/lex.name#3.1)

Reviewer 2: ```suggestion
   begin with an underscore and a capital letter, as per [\[lex.name\]/3.1][].
```

2. Your changes don't introduce change ABI for any type or function that a user may use (including adding or removing
  non-static data members, adding or removing virtual member functions, changing whether a type is an aggregate
  or trivially copyable, etc.).

3. Your changes are written from scratch using only acceptable sources:
   * this repository,
   * the C++ Working Draft (including any cited standards),
   * other WG21 papers (excluding reference implementations outside of proposed standard wording),

>    * LWG issues, and

Reviewer 2: ```suggestion
   * LWG issues, or
```

>    * a project listed in [NOTICE.txt][NOTICE.txt] (make sure to cite the project in the PR description!).

Reviewer 2: ```suggestion
   * a project listed in [NOTICE.txt][] (make sure to cite the project in the PR description!).
```

>   If your changes are derived from any other project, you _must_ mention it
>   in the pull request description, so we can determine whether the license is compatible and whether any other
>   steps need to be taken.

Reviewer 2: ```suggestion
If your changes are derived from any other project, you _must_ mention it
in the pull request description, so we can determine whether the license is compatible and whether any other
steps need to be taken.
```

# Code of Conduct

This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or
contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

[label:"good first issue"]:
   https://github.com/microsoft/STL/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22
[label:"high priority"]: https://github.com/microsoft/STL/issues?q=is%3Aopen+is%3Aissue+label%3A%22high+priority%22
[label:"help wanted"]: https://github.com/microsoft/STL/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22
[review:changes-requested]: https://github.com/microsoft/STL/pulls?q=is%3Apr+is%3Aopen+review%3Achanges-requested
[wiki:advice-for-reviewing]: https://github.com/microsoft/STL/wiki/Advice-for-Reviewing-Pull-Requests
[NOTICE.txt]: https://github.com/microsoft/STL/blob/main/NOTICE.txt

> [wiki:videos]: https://github.com/microsoft/STL/wiki/Code-Review-Videos

Reviewer 2: ```suggestion
[wiki:videos]: https://github.com/microsoft/STL/wiki/Code-Review-Videos
[\[lex.name\]/3.1]: https://eel.is/c++draft/lex.name#3.1
```
